aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -143,6 +143,53 @@ Tensor where(const Tensor& condition, const Tensor& self, const Tensor& other) {
   return at::_s_where(b_condition, b_self, b_other);
 }
 
+Tensor where(const Tensor& condition, Scalar self, const Tensor& other) {
```

Comment:
Instead of doing this, why not just wrap the scalar into a tensor?

Reply:
Could you please give an example. Did not understand exactly.
Thanks.

Reply:
Take a look at the code reference for add in the comment below, does that make more sense?

Reply:
https://github.com/pytorch/pytorch/blob/80d5b3785b88f83eb598e393d0137f045b979c4b/aten/src/ATen/native/BinaryOps.cpp#L353-L357

Which calls `scalar_to_tensor` (snippet below) which creates a tensor based on the general type of the scalar which I feel 
is rigid.
https://github.com/pytorch/pytorch/blob/80d5b3785b88f83eb598e393d0137f045b979c4b/aten/src/ATen/ScalarOps.h#L12-L36

Using this directly we won't be able to 
```python
torch.where(a > 0.5, a, 0) 
```
as 0 will be converted to a `LongTensor` and `where` will fail with type mismatch.

Instead inspired by this we can do something like the following ( note: haven't actually tried it)
```

inline at::Tensor scalar_to_tensor_with_options(Scalar s, const TensorOptions options) { 
   return at::native::scalar_tensor(s, options);
}

 static Tensor wrapped_scalar_tensor(Scalar scalar, TensorOptions options) { 
   auto tensor = scalar_to_tensor_with_options(scalar, options); 
   tensor.unsafeGetTensorImpl()->set_wrapped_number(true); 
   return tensor; 
 } 
```

Do let me know what you feel about it.
Thanks!

Reply:
I like your idea but I think you're better off implementing type promotion for where, so you pass the tensors in as long tensor or double tensor and then type promote. This has two advantages:

- it will work on tensor inputs
- you won't cast doubles to long if a is long and someone passes a value like 1.5

What are your thoughts?

Reply:
Agree that type promotion would be more elegant.
But my plan was to work on type promotion post this PR as that would totally change the scope of this PR. 

Adding type promotion will also change the behaviour of current implementation and will need its own set of tests with the ones added in this so I was refraining to do it in this PR.

Reply:
That's a great point and I think it's separable, but I would still implement this PR, as just wrapping these scalars because it's on path to that future work and an improvement over today.

Reply:
Sure. Will try to implement something like the above snippet.
Thanks.

Reply:
Cool, but no `scalar_to_tensor_with_options`, right? It's a cool idea but you'd have to remove it later.

Reply:
I was actually planning with `scalar_to_tensor_with_options` .
 Because if we use current `scalar_to_tensor`,
https://github.com/pytorch/pytorch/blob/80d5b3785b88f83eb598e393d0137f045b979c4b/aten/src/ATen/ScalarOps.h#L12-L36
We will only be able to use the tensor with `dtype=torch.double` when the scalar is floating type, as `where` only allows tensors of same dtype.
```python
torch.where(condition, x.to(torch.double), 2.) # Valid
torch.where(condition, x.to(torch.float), 2.) # Invalid as 2. will be wrapped as a DoubleTensor
```

Similar will be the case when using a int scalar, only `dtype=torch.long` will be valid for the corresponding tensor input.
This seems very inconvenient. 

So for now I guess `scalar_to_tensor_with_options` is not so bad. I'll add it just above `where` implementation in anonymous namespace and leave a note to remove it once type promotion is added.

Do let me know how you feel about this.

Reply:
I understand your point and think that limitation is OK for now. Otherwise you'll have very odd behavior. Like two scalars won't work, and a long tensor + a float scalar will "work" but do the wrong thing.

Reply:
Sure will use the existing code.

Agree that behaviour will be weird when scalars will be demoted.